casper/network.py

```diff
@@ -10,74 +11,62 @@ def __init__(self, validator_set, protocol=BlockchainProtocol, force_justify_mes
             self._collect_initial_messages(),
             protocol.initial_message(None)
         )
+        self.message_queues = {
+            validator: PriorityQueue()
+            for validator in self.validator_set
+        }
+        self.time = 0
 
         self.force_justify_messages = force_justify_messages
 
-    def propagate_message_to_validator(self, message, validator):
-        """Propagate a message to a validator."""
-        assert message.hash in self.global_view.justified_messages, (
-            "...expected only to propagate messages "
-            "from the global view")
-        assert validator in self.validator_set, "...expected a known validator"
-
-        validator.receive_messages(set([message]))
-
-        # HACK TO SEND ALL MESSAGES BEFORE NETWORK REWORK
-        if self.force_justify_messages:
-            self._propagate_messages_needed_to_justify(message, validator)
-
-    def get_message_from_validator(self, validator):
-        """Get a message from a validator."""
-        assert validator in self.validator_set, "...expected a known validator"
-
-        new_message = validator.make_new_message()
-        self.global_view.add_messages(set([new_message]))
-        assert new_message.hash in self.global_view.justified_messages
-
-        return new_message
-
-    def view_initialization(self, view):
-        """Initalizes all validators with all messages in some view."""
-        messages = view.justified_messages.values()
-
-        self.global_view.add_messages(messages)
-
+    def delay(self, sender, receiver):
+        '''Must be defined in child class.
+        Returns delay of next message for sender to receiver'''
+        raise NotImplementedError
+
+    #
+    # Validator API to Network
+    #
+    def send(self, validator, message):
+        self.global_view.add_messages(
+            set([message])
+        )
+        self.message_queues[validator].put((
+            self.time + self.delay(message.sender, validator),
+            message
+        ))
+
+    def receive(self, validator):
+        queue = self.message_queues[validator]
+        if queue.qsize() == 0:
+            return None
+        if queue.peek()[0] > self.time:
+            return None
+
+        return queue.get()[1]
+
+    def receive_all(self, validator):
```

Comment:
This function seems a bit misnamed w/ `send_to_all` below. It seems like `receive_all` should have all validators receive their current messages - rather than a single validator receiving all of their current messages.

Maybe change the semantics of receive so it returns a set of current messages? Then, have `receive_all` return a dictionary to this for all validators? Just so the relationship between send and send_to_all and receive and receive_to_all are the same.

Not sure if there are further reaching implications of this change, but at the very least I might rename to make it more clear.

Reply:
@naterush I agree that it would be nice to have a semantic distinction between the two. Perhaps `receive_own` vs `receive_all`...

But i am wondering wether it is contrary to the networking model to have another `receive_all` fn that is able to receive all of the other validators' priority queues as well. (unless for pure modeling purposes)

I do see how such a comprehensive `receive_all` could be used in a situation where we want to instantly synchronize a validators view with all of the messages floating in the network. Or iterate through all of the validators to fully synchronize the whole network.

Reply:
I don't think `receive_all` in the sense that it would pop them all out of all of the queues should be in the network model. I think validators should only be able to pull their messages off the queue (assuming honest calls or some sort of access control).

I'm down to change the name to something clearer. `receive_own` seems too close to what `receive` is actually doing. maybe `receive_available` or `receive_all_available`. Or maybe `receive` should just always return a list -- empty if nothing available, otherwise filled with as many things that are available. I created `receive` as is because of the networking paper, but having return a list doesn't seem like an issue to me.

thoughts?

Reply:
Renaming receive_all_available or changing receive both seem good. Maybe best to just rename to receive_all_available, so we can have higher fidelity with controlling when validators receive messages? 

Whatever you think is best :)

Reply:
`receive_all_available` for now